File: csharp/0701-insert-into-a-binary-search-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution
{
    public TreeNode InsertIntoBST(TreeNode root, int val)
    {
        if (root is null) return new TreeNode(val);

        TreeNode cur = root;

        while (cur is not null)
        {
            if (cur.val < val)
            {
                if (cur.right is null)
                {
                    cur.right = new TreeNode(val);
                    break;
                }

                cur = cur.right;
                continue;
            }

            if (cur.val > val)
            {
                if (cur.left is null)
                {
                    cur.left = new TreeNode(val);
                    break;
                }

                cur = cur.left;
                continue;
            }
        }

        return root;
    }
}

File: cpp/0701-insert-into-a-binary-search-tree.cpp
/*
    You are given the root node of a binary search tree (BST) and a value to insert into the tree. 
    Return the root node of the BST after the insertion. It is guaranteed that the new value does not exist in the original BST.

    Notice that there may exist multiple valid ways for the insertion, as long as the tree remains a BST after insertion. 
    You can return any of them.

    Ex. Input: root = [4,2,7,1,3], val = 5
        Output: [4,2,7,1,3,5]

    Time  : O(N)
    Space : O(N)
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode * create(int val) {
        TreeNode *n = new TreeNode;
        n -> val = val;
        n -> left = n -> right = NULL;
        return n;
    }

    TreeNode* insertIntoBST(TreeNode* root, int val) {
        if(root == NULL) 
		    return create(val);

        else if(val > root -> val)
		    root -> right = insertIntoBST(root -> right, val); 
        else if(val < root -> val)
		    root -> left = insertIntoBST(root -> left, val);
	    return root;
    }
};
